src/daft-io/src/google_cloud.rs

```diff
@@ -334,9 +336,49 @@ pub(crate) struct GCSSource {
     client: GCSClientWrapper,
 }
 
+#[derive(Debug, Clone)]
+struct FixedTokenSource {
+    token: String,
+}
+
+impl TokenSourceProvider for FixedTokenSource {
+    fn token_source(&self) -> Arc<dyn TokenSource> {
+        Arc::new(self.clone())
+    }
+}
+
+#[async_trait]
+impl TokenSource for FixedTokenSource {
+    async fn token(&self) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
+        Ok(format!("Bearer {0}", self.token))
+    }
+}
+
 impl GCSSource {
     pub async fn get_client(config: &GCSConfig) -> super::Result<Arc<Self>> {
-        let config = if !config.anonymous {
+        let mut client_config = if config.anonymous {
+            ClientConfig::default().anonymous()
+        } else if let Some(creds) = &config.credentials {
+            let creds_file = CredentialsFile::new_from_file(creds.clone()).await;
```

Comment:
These functions may have side effects so i'd recommend doing one and checking the result and then the other if needed. For sync code there's [`ok_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else)

Reply:
`ok_or_else` doesn't work because I need to await the second result, but I'll rewrite it so that it tries file first with a match